pymeasure/instruments/anritsu/anritsuMS2090A.py

```diff
@@ -0,0 +1,535 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2022 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+import logging
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import (
+    strict_discrete_set,
+    truncated_range
+)
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class AnritsuMS2090A(Instrument):
+    """Anritsu MS2090A Handheld Spectrum Analyzer."""
+
+    def __init__(self, adapter, name="Anritsu MS2090A Handheld Spectrum Analyzer", **kwargs):
+        self.analysis_mode = None
+        super().__init__(
+            adapter, name, **kwargs)
+
+    def double_validation_value_and_freq(value, values):
+        pass_discrete = None
+        pass_range = None
+
+        value_2 = value[0]
+        value_1 = value[1]
+        values_2 = values[0]
+        values_1 = values[1]
+
+        if value_1 in values_1:
+            pass_discrete = value_1
+        else:
+            raise ValueError('Value of {:s} is not in the discrete set {:s}'.format(
+                value_1, values_1
+            ))
+
+        if min(values_2) <= value_2 <= max(values_2):
+            pass_range = value_2
+        else:
+            raise ValueError('Value of {:g} is not in range [{:g},{:g}]'.format(
+                value_2, min(values_2), max(values_2)
+            ))
+        return f'{pass_range} {pass_discrete}'

```

Comment:
The validator has to be outside the class definition.

Don't you think it is cumbersome to write `inst.frequency = 500, "MHz"` instead of `inst.frequency = 500e6`, especially if you try to access these controls from a program?

I think the best way is, if the methods accept a value in Hz and return the value in Hz, such that nobody has to fiddle with the units string.

If for some reason (accuracy etc.) you have to use the different units, you could make your validator accept a float (frequency in Hz) and then the validator decides what unit (Hz, kHz, MHz, or GHz) is best and hands the (modified) float and the corresponding unit string to the device.

Reply:
Yes, maybe it is better only introduce it on Hz. I am going to change it